### ISSUE TITLE
Websocket connection is not closed, when tcp client disconnects

### ISSUE THREAD

**Issue by toewsar** [SoftwareAG/cumulocity-remote-access-local-proxy#38]
Some protocols do not keep the connection open. And if they open again they assume to have a clean session. 
But because the c8ylp does not disconnect the websocket, the tcp connection on the device side remains open. 

Therefore the c8ylp have to close the websocket conenction, if the tcp client disconnects. This would notify the device that the connection is closed and the device would then disconnect the tcp server, too.

I would be happy if you could fix this.

**Comment by reubenmiller** [comment]
Thanks @toewsar for the ticket. To make the change easier (and to add a test case for it), would you be able to answer the following questions?

1. I think this behaviour changed with the migration from v1 to v2. Could you please verify if it works for you in v1.5.0, and then stops working in v2?
2. Is there an example server/client that we could use to check the handling of this (and use to verify such a change in behaviour)?

**Comment by toewsar** [comment]
Yes, you can test it. I tried to tunnel the CODESYS IDE <-> CODESYS Gateway. 
You need to install [CODESYS Development System V3](https://store.codesys.com/de/codesys.html) and on the device you need to install the [CODESYS Edge Gateway for Linux](https://store.codesys.com/de/codesys-edge-gateway-for-linux.html).

I'll try to test it today with 1.5.0.

**Comment by switschel** [comment]
@toewsar This depends in which mode you start the c8ylp. In server mode the c8ylp acts like a server so you have to shutdown it manually to terminate the websocket connection. So if the TCP-Client disconnects you have to make sure to also terminate the c8ylp server if this is desired.

@reubenmiller In 1.5 we had the script mode doing exactly that: Terminate c8ylp when tcp client disconnects. Not sure if we should add a flag to server or we can achieve that with connect command.

**Comment by reubenmiller** [comment]
Or using the in-built "command" plugin

```sh
c8ylp plugin command --env-file .env --reconnects 1 device01 './runsomething $C8Y_PORT'
```

**Comment by switschel** [comment]
@reubenmiller reconnect 0 will set it to infinite reconnects
``
  --reconnects INTEGER RANGE  number of reconnects to the Cloud Remote
                              Service. 0 for infinite reconnects  [env var:
                              C8YLP_RECONNECTS; default: 5; -1<=x<=10] 
``
Maybe we can use -1 to disable any reconnect and shutdown the c8ylp after first websocket connection?

**Comment by reubenmiller** [comment]
@switschel Yes that sounds reasonable. Though I still want to make sure that this is the actual issue though, and not some other internal workings with the connection/disconnection of the sockets.

**Comment by switschel** [comment]
Btw. it is already implemented: https://github.com/SoftwareAG/cumulocity-remote-access-local-proxy/blob/27b7781df2342105f3aa90e2eee6f8d1664a3b88/c8ylp/websocket_client/ws_client.py#L177

@toewsar Can you try if adding '--reconnect -1' is the behavior you expect to have?

**Comment by reubenmiller** [comment]
> Btw. it is already implemented:
> 
> https://github.com/SoftwareAG/cumulocity-remote-access-local-proxy/blob/27b7781df2342105f3aa90e2eee6f8d1664a3b88/c8ylp/websocket_client/ws_client.py#L177
> 
> @toewsar Can you try if adding '--reconnects -1' is the behavior you expect to have?

That is the right spot, however the logic is inverted. It should be (self.max_retries == -1), as the if statement is to decide if the connection should be terminated. I have made a change on this [fix/disable-reconnects](https://github.com/SoftwareAG/cumulocity-remote-access-local-proxy/tree/fix/disable-reconnects) branch

**Comment by toewsar** [comment]
I tried, but without success. The problem ist, that the tcp-server does not forward the close to the websocket client. 
See https://github.com/SoftwareAG/cumulocity-remote-access-local-proxy/blob/27b7781df2342105f3aa90e2eee6f8d1664a3b88/c8ylp/tcp_socket/tcp_server.py#L81

**Comment by reubenmiller** [comment]
@switschel I've updated the PR #40 to hide the reconnects option and print out a warning that it is deprecated (so that we don't break any existing scripts).

**Comment by reubenmiller** [comment]
@toewsar This has been fixed and released in v2.1.1
